# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [19]:
import numpy as np
from scipy import stats

import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

In [18]:
mu0 = 123.7
alpha_1 = 0.10
times = np.array([128, 135, 121, 142, 126, 151, 123])

# Завдання 1
print('#1')
print('ГІПОТЕЗИ:')
print('H₀: μ = 123.7')
print('H₁: μ > 123.7')
print('Тип тесту: правосторонній')
print()

# Завдання 2
print('#2')
x_bar = np.mean(times)
s = np.std(times, ddof=1)
n = len(times)
print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"К-сть вибіркових даних: {n}")
print()

# Завдання 3
print('#3')
res = stats.ttest_1samp(times, popmean=mu0, alternative="greater")
t_crit = stats.t.ppf(1 - alpha_1, res.df)
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.3f}")
print(f"t_crit = {t_crit:.3f}")
print()

# Завдання 4
print('#4')
if res.pvalue < alpha_1:
    print(f"Відхиляємо H0: середній час до покупки після редизайну сайту більший за 123.7 хвилин на рівні значущості {alpha_1}")
else:
    print("Не відхиляємо H0: доказів недостатньо")
print()

# Завдання 5
print('#5')
print('На основі тесту ми можемо чітко ствердити, що редизайн сайту статистично значущо подовжив час до покупки користувачів')
print()

# Завдання 6
print('#6')
alpha_2 = 0.05
print(f"РІВЕНЬ ЗНАЧУЩОСТІ: {alpha_2}")
res = stats.ttest_1samp(times, popmean=mu0, alternative="greater")
t_crit = stats.t.ppf(1 - alpha_2, res.df)
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.3f}")
print(f"t_crit = {t_crit:.3f}")
print()

if res.pvalue < alpha_2:
    print(f"Відхиляємо H0: середній час до покупки після редизайну сайту більший за 123.7 хвилин на рівні значущості {alpha_2}")
else:
    print("Не відхиляємо H0: доказів недостатньо")
print()

print(f"При рівні значущості: {alpha_2} рішення щодо Відхилення H0 не змінилося")

#1
ГІПОТЕЗИ:
H₀: μ = 123.7
H₁: μ > 123.7
Тип тесту: правосторонній

#2
Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
К-сть вибіркових даних: 7

#3
scipy: t = 2.069, p-value = 0.042
t_crit = 1.440

#4
Відхиляємо H0: середній час до покупки після редизайну сайту більший за 123.7 хвилин на рівні значущості 0.1

#5
На основі тесту ми можемо чітко ствердити, що редизайн сайту статистично значущо подовжив час до покупки користувачів

#6
РІВЕНЬ ЗНАЧУЩОСТІ: 0.05
scipy: t = 2.069, p-value = 0.042
t_crit = 1.943

Відхиляємо H0: середній час до покупки після редизайну сайту більший за 123.7 хвилин на рівні значущості 0.05

При рівні значущості: 0.05 рішення щодо Відхилення H0 не змінилося


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [24]:
count = 311
nobs = 1200
p0 = 0.23
alpha = 0.03

print('ГІПОТЕЗИ:')
print('H₀: p = 0.23')
print('H₁: p > 0.23')
print('Тип тесту: правосторонній')
print()

stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha, method='normal')

print("Тест")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")
print()

if pval < alpha:
    print(f"При рівні значущості: {alpha} ми відхиляємо H₀. Також можемо стверджувати що впізнаваність логотипу статистично значущо зросла та тепер становить більше ніж 23%")
else:
    print(f"При рівні значущості: {alpha} ми НЕ ВІДХИЛЯЄМО H₀, адже нас немає підстав стверджувати, що впізнаваність логотипу зросла")

ГІПОТЕЗИ:
H₀: p = 0.23
H₁: p > 0.23
Тип тесту: правосторонній

Тест
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)

При рівні значущості: 0.03 ми відхиляємо H₀. Також можемо стверджувати що впізнаваність логотипу статистично значущо зросла та тепер становить більше ніж 23%


З обрахунків, ми бачимо, що p-value = 0.0106, а це менше за рівень значущості 0.03, то ми відхиляємо H0.
Рекламна кампанія була успішною, і впізнаваність логотипу зросла.

Довірчий інтервал в нас складає (23.171%, 28.661%) - тобто умова великої вибірки виконується [0, 1] та це підтверджує цей висновок, оскільки навіть його мінімальне прогнозоване значення перевищує початкові 23%.